src/main/java/com/amazon/opendistroforelasticsearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -158,106 +152,165 @@ private static final boolean isLocalAll(final String... requestedPatterns) {
         return false;
     }
 
-    private Resolved resolveIndexPatterns(final IndicesOptions indicesOptions, final boolean enableCrossClusterResolution, final String... requestedPatterns0) {
+    private class ResolvedIndicesProvider implements IndicesProvider {
+        private final ImmutableSet.Builder<String> aliases;
+        private final ImmutableSet.Builder<String> allIndices;
+        private final ImmutableSet.Builder<String> originalRequested;
+        private final ImmutableSet.Builder<String> remoteIndices;
+        // set of previously resolved index requests to avoid resolving
+        // the same index more than once while processing bulk requests
+        private final Set<MultiKey> alreadyResolved;
```

Comment:
Why we use `MultiKey` here?

Reply:
As it provides the required functionality without any additional dependencies and simplifies the code.